### PR TITLE
Recursive conditional types

### LINKED ISSUE #26223: Add a generic way to specify length of a tuple type
## Search Terms
tuple, type, length

## Suggestion

I'd like to see a way to specify the length of a tuple or array type when declaring it. For short tuples, it may be viable to write them out by hand like `[number, number]`, but whenever the length is longer (say 10+), it becomes a chore.

Using the following approach to define a fixed-length tuple
```ts
// @ts-ignore interface can only extend interface or class
interface FixedLengthArray<T extends any[], L extends number> extends T {
	length: L;
}
```
has a few drawbacks:
1.: we cannot extend T because its not an interface or class.  
2.: when ignoring that error, tuple literals are not inferred as tuples but as arrays when assigning them:
```ts
type Foo = FixedLengthArray<number[], 5>;
const foo1: Foo = [1, 2, 3, 4, 5]; // error: length is not compatible, number not assignable to 5
const foo2: Foo = [1, 2, 3, 4, 5] as [1, 2, 3, 4, 5]; // works, but defeats the purpose of that fixed-length
```
On the other hand, when using manually-written tuples, this works:
```ts
type Bar = [number, number, number, number, number];
const bar1: Bar = [1, 2, 3, 4, 5]; // no error
```

## Use Cases
* Declaring fixed-length tuples or arrays with a large number of items
* Creating a Range union (e.g. `0 | 1 | 2 | 3 | 4` from the keys of such a type) without having to resort to recursive types (currently broken, see #26155)

## Examples
I'll leave the syntax up for discussion, and provide the length in the square brackets for now:
```ts
type T1 = number[3]; // equals [number, number, number]
type K1 = keyof T1; // 0 | 1 | 2
type T2 = number[100]; // equals [number, number, ..., number] (100 items total)
type K2 = keyof T2; // 0 | 1 | ... | 99
type T3 = any[0]; // equals [ ], type annotation is somewhat obsolete here
type K3 = keyof T3; // never
```
The basic rules for tuple assignability should apply, as if the types were written out by hand.

## Related:
https://github.com/Microsoft/TypeScript/issues/24350
https://github.com/Microsoft/TypeScript/issues/18471

## Checklist

My suggestion meets these guidelines:
* [x] This wouldn't be a breaking change in existing TypeScript / JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. new expression-level syntax)

### LINKED ISSUE #26980: Feature request: lift the circular constraint for conditional types
## Search Terms

circular type conditional type

## Suggestion

Currently, `Last1` is valid, but seemingly-equivalent `Last2` is not:

```ts
type Head<T extends any[]> = T extends [infer X, ...any[]] ? X : never;

type Tail<T extends any[]> =
    ((...x: T) => void) extends ((x: any, ...xs: infer XS) => void) ? XS : never;

type Last1<T extends any[]> = {
  0: never,
  1: Head<T>,
  2: Last1<Tail<T>>,
}[T extends [] ? 0 : T extends [any] ? 1 : 2];

type Last2<T extends any[]> =
    T extends [] ? never :
    T extends [infer R] ? R :
    Last2<Tail<T>>;
```

My suggestion is to make `Last2` valid, since the recursion isn't necessarily unbounded.

## Use Cases

It'd make recursive conditional types a lot easier and more intuitive to write. It'd also eliminate the need to guard against impossible cases when using the workaround like when using `Last1` above. If you wanted to mandate termination, all I'd want is direct recursion - this makes it much easier to check for infinite recursion. (Other types are generally assumed to terminate, so you're only assessing control flow.)

## Examples

See above in the suggestion summary. Here's a concrete example of how [this file](https://github.com/kgtkr/typepark/blob/master/src/pipe.ts) (with [dependency](https://github.com/kgtkr/typepark/blob/master/src/list.ts)) would be simplified:

```ts
type Last<L extends any[], D = never> =
    L extends [] ? D :
    L extends [infer H] ? H :
    ((...l: L) => any) extends ((h: any, ...t: infer T) => any) ? Last<T> :
    D;

type Append<T extends any[], H> =
    ((h: H, ...t: T) => any) extends ((...l: infer L) => any) ? L : never;

type Reverse<L extends any[], R extends any[] = []> =
    ((...l: L) => any) extends ((h: infer H, ...t: infer T) => any) ?
        Reverse<T, Append<R, H>> :
        R;

type Compose<L extends any[], V, R extends any[] = []> =
    ((...l: L) => any) extends ((a: infer H, ...t: infer T) => any) ?
        Compose<T, H, Append<R, (x: V) => H>> :
        R;

export type PipeFunc<T extends any[], V> =
    (...f: Reverse<Compose<T, V>>) => ((x: V) => Last<T, V>);
```

Currently, you could achieve similar via [this](https://www.typescriptlang.org/play/index.html#src=type%20Last%3CL%20extends%20any%5B%5D%2C%20D%20%3D%20never%3E%20%3D%20%7B%0D%0A%20%20%20%200%3A%20D%2C%0D%0A%20%20%20%201%3A%20L%20extends%20%5Binfer%20H%5D%20%3F%20H%20%3A%20never%2C%0D%0A%20%20%20%202%3A%20((...l%3A%20L)%20%3D%3E%20any)%20extends%20((h%3A%20any%2C%20...t%3A%20infer%20T)%20%3D%3E%20any)%20%3F%20Last%3CT%3E%20%3A%20D%2C%0D%0A%7D%5BL%20extends%20%5B%5D%20%3F%200%20%3A%20L%20extends%20%5Bany%5D%20%3F%201%20%3A%202%5D%3B%0D%0A%0D%0Atype%20Append%3CT%20extends%20any%5B%5D%2C%20H%3E%20%3D%0D%0A%20%20%20%20((h%3A%20H%2C%20...t%3A%20T)%20%3D%3E%20any)%20extends%20((...l%3A%20infer%20L)%20%3D%3E%20any)%20%3F%20L%20%3A%20never%3B%0D%0A%0D%0Atype%20Reverse%3CL%20extends%20any%5B%5D%2C%20R%20extends%20any%5B%5D%20%3D%20%5B%5D%3E%20%3D%20%7B%0D%0A%20%20%20%200%3A%20R%2C%0D%0A%20%20%20%201%3A%20((...l%3A%20L)%20%3D%3E%20any)%20extends%20((h%3A%20infer%20H%2C%20...t%3A%20infer%20T)%20%3D%3E%20any)%20%3F%0D%0A%20%20%20%20%20%20%20%20Reverse%3CT%2C%20Append%3CR%2C%20H%3E%3E%20%3A%0D%0A%20%20%20%20%20%20%20%20never%2C%0D%0A%7D%5BL%20extends%20%5Bany%2C%20...any%5B%5D%5D%20%3F%201%20%3A%200%5D%3B%0D%0A%0D%0Atype%20Compose%3CL%20extends%20any%5B%5D%2C%20V%2C%20R%20extends%20any%5B%5D%20%3D%20%5B%5D%3E%20%3D%20%7B%0D%0A%20%20%20%200%3A%20R%2C%0D%0A%20%20%20%201%3A%20((...l%3A%20L)%20%3D%3E%20any)%20extends%20((a%3A%20infer%20H%2C%20...t%3A%20infer%20T)%20%3D%3E%20any)%20%3F%0D%0A%20%20%20%20%20%20%20%20Compose%3CT%2C%20H%2C%20Append%3CR%2C%20(x%3A%20V)%20%3D%3E%20H%3E%3E%0D%0A%20%20%20%20%20%20%20%20%3A%20never%2C%0D%0A%7D%5BL%20extends%20%5Bany%2C%20...any%5B%5D%5D%20%3F%201%20%3A%200%5D%3B%0D%0A%0D%0Aexport%20type%20PipeFunc%3CT%20extends%20any%5B%5D%2C%20V%3E%20%3D%0D%0A%20%20%20%20(...f%3A%20Reverse%3CCompose%3CT%2C%20V%3E%3E)%20%3D%3E%20((x%3A%20V)%20%3D%3E%20Last%3CT%2C%20V%3E)%3B), but as you can see, it's highly repetitive and boilerplatey:

```ts
type Last<L extends any[], D = never> = {
    0: D,
    1: L extends [infer H] ? H : never,
    2: ((...l: L) => any) extends ((h: any, ...t: infer T) => any) ? Last<T> : D,
}[L extends [] ? 0 : L extends [any] ? 1 : 2];

type Append<T extends any[], H> =
    ((h: H, ...t: T) => any) extends ((...l: infer L) => any) ? L : never;

type Reverse<L extends any[], R extends any[] = []> = {
    0: R,
    1: ((...l: L) => any) extends ((h: infer H, ...t: infer T) => any) ?
        Reverse<T, Append<R, H>> :
        never,
}[L extends [any, ...any[]] ? 1 : 0];

type Compose<L extends any[], V, R extends any[] = []> = {
    0: R,
    1: ((...l: L) => any) extends ((a: infer H, ...t: infer T) => any) ?
        Compose<T, H, Append<R, (x: V) => H>>
        : never,
}[L extends [any, ...any[]] ? 1 : 0];

export type PipeFunc<T extends any[], V> =
    (...f: Reverse<Compose<T, V>>) => ((x: V) => Last<T, V>);
```

(The original code snippet is linked to from [here](https://github.com/Microsoft/TypeScript/issues/5453#issuecomment-419636447) as a possible solution to the issue of `_.compose`, `_.flow`, and friends.)

## Checklist

My suggestion meets these guidelines:
* [x] This wouldn't be a breaking change in existing TypeScript / JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. new expression-level syntax)

### LINKED ISSUE #37801: Type alias recursion is resolved eagerly when the alias is generic and its arguments are derived from inferred members of conditional types
**TypeScript Version:**  3.9.0-beta

**Search Terms:** recursive, recursive generic

**Code**

```
type Intersect<U extends any[], R = unknown> =
    ((...u: U) => any) extends ((h: infer H, ...t: infer T) => any)
        ? Intersect<T, R & H>
        : R

const value: Intersect<[{a: number}, {b: number}]> = {
    a: 1,
    b: 2
}
```

*Note: the usual workaround of an immediately-indexed intermediate object type still works.*

**Expected behavior:** It to type check without warnings

**Actual behavior:**

- At line 1: `Type alias 'Intersect' circularly references itself. (2456)`
- At line 3: `Type 'Intersect' is not generic. (2315)`
- At line 6: `Type 'Intersect' is not generic. (2315)`

**Playground Link:**  [Link](https://www.typescriptlang.org/play?ts=3.9.0-beta#code/C4TwDgpgBAkgdsCAnAzhAxsAPAVShAD0TgBMUoBDOEAbQF0AaKAJSgF4oBXOAazgHsA7nAB87AFBQpUABQyAdIs4AuKDgCU7MVRCbCxMrJkALVQEs4AM2RQAEk0Xzg5qzYAqmtturrJ0-wD8sAjIaJhYbkysAGR2In7+UqrM4uLo-HAowFAAbhQANpwQqvCIqBjYNADeFKpwnAC2AEbIAL5MVU11jS1IrXRiHFUJlKoAjAwjXVAATOKtQA)

**Related Issues:** 

- #37426 (root cause might be this)
- #37344
- Probably about 25%-30% of all other recursive type issues.
- My use case is basically solving #26058 in userland, and doing [this](https://stackoverflow.com/questions/50374908/transform-union-type-to-intersection-type/50375286#50375286) with `SomeTuple[number]` won't work because I don't want to recurse.

### PR DESCRIPTION
**Pull request by ahejlsberg**
With this PR we officially support recursive conditional types. For example:

```ts
// Awaiting promises

type Awaited<T> =
    T extends null | undefined ? T :
    T extends PromiseLike<infer U> ? Awaited<U> :
    T;

type P1 = Awaited<Promise<string>>;  // string
type P2 = Awaited<Promise<Promise<string>>>;  // string
type P3 = Awaited<Promise<string | Promise<Promise<number> | undefined>>;  // string | number | undefined
```

A bit of context... When conditional types were first introduced, we explicitly restricted them to be non-recursive, i.e. we made it an error for a conditional type to directly or indirectly reference itself. This restriction was put in place primarily as a safeguard against runaway infinite recursion which the compiler didn't handle well at the time. However, it turns out it was still possible to construct recursive conditionally resolved types by combining object types (which have deferred resolution of property types) and conditional types using indexed access types (see #14833 for the core idea). In spite of being cumbersome and non-intuitive, this trick has become commonplace in several libraries. Consequently, over time we have "hardened" the compiler against infinite recursion with depth limiters in relationships, type inference, type instantiation, constraint computation, and so on. We're now at a point where it seems reasonable support an intuitive way of writing recursive conditional types.

Some more examples:

```ts
// Flattening arrays

type Flatten<T extends readonly unknown[]> = T extends unknown[] ? _Flatten<T>[] : readonly _Flatten<T>[];
type _Flatten<T> = T extends readonly (infer U)[] ? _Flatten<U> : T;

type InfiniteArray<T> = InfiniteArray<T>[];

type A1 = Flatten<string[][][]>;  // string[]
type A2 = Flatten<string[][] | readonly (number[] | boolean[][])[]>;  // string[] | readonly (number | boolean)[]
type A3 = Flatten<InfiniteArray<string>>;
type A4 = A3[0];  // Infinite depth error

// Repeating tuples

type TupleOf<T, N extends number> = N extends N ? number extends N ? T[] : _TupleOf<T, N, []> : never;
type _TupleOf<T, N extends number, R extends unknown[]> = R['length'] extends N ? R : _TupleOf<T, N, [T, ...R]>;

type T1 = TupleOf<string, 3>;  // [string, string, string]
type T2 = TupleOf<number, 0 | 2 | 4>;  // [] | [number, number] | [number, number, number, number]
type T3 = TupleOf<number, number>;  // number[]
type T4 = TupleOf<number, 100>;  // Depth error
```

Note that this PR doesn't change the recursion depth limits that are already in place. For example, an error is reported on `T4` above because its resolution exceeds the limit of 50 nested type instantiations.

The type inference streamlining contained in the PR fixes several issues with inference to recursive types. For example:

```ts
interface Box<T> { value: T };
type RecBox<T> = T | Box<RecBox<T>>;

declare function unbox<T>(box: RecBox<T>): T

type T1 = Box<string>;
type T2 = Box<T1>;
type T3 = Box<T2>;
type T4 = Box<T3>;
type T5 = Box<T4>;
type T6 = Box<T5>;

declare let b1: Box<Box<Box<Box<Box<Box<string>>>>>>;
declare let b2: T6;

unbox(b1);  // string
unbox(b2);  // string (previously T6)
unbox({ value: { value: { value: 5 }}});  // number (previously { value: { value: number }})
```

Fixes #26223.
Fixes #26980.
Fixes #37801.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test this

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -19533,7 +19494,7 @@ namespace ts {
                 priority = savePriority;
             }
 
-            function invokeOnce(source: Type, target: Type, action: (source: Type, target: Type) => void) {
+            function invokeWithDepthLimit(source: Type, target: Type, action: (source: Type, target: Type) => void) {
```
Comment (weswigham) [inline src/compiler/checker.ts:19497]: This is pretty much the inference equivalent of `recursiveTypeRelatedTo` at this point.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -18199,56 +18186,44 @@ namespace ts {
...
+            if (type.flags & TypeFlags.IndexedAccess) {
+                // Identity is the leftmost object type in a chain of indexed accesses, eg, in A[P][Q] it is A
+                do {
+                    type = (type as IndexedAccessType).objectType;
+                } while (type.flags & TypeFlags.IndexedAccess);
+                return type;
+            }
+            if (type.flags & TypeFlags.Conditional) {
+                // The root object represents the origin of the conditional type
+                return (type as ConditionalType).root;
```
Comment (weswigham) [inline src/compiler/checker.ts:18224]: Huh, and this doesn't affect the `user` baselines or DT?

**Comment by treybrisbane** [comment]
Awesome to finally see this! 😀

**Comment by DanielRosenwasser** [comment]
@ahejlsberg maybe something worthwhile is to build another PR on top of this to see if changing the definition of [`FlatArray`](https://github.com/microsoft/TypeScript/blob/1f5caf554c4c468767097db36cea52e1911e2ce2/lib/lib.es2019.array.d.ts#L21) would impact real-world code.

**Comment by ahejlsberg** [comment]
The tests revealed OOMs in a few projects due to the switch to use `isDeeplyNestedType` for recursion tracking in type inference (which permits up to five levels of recursion). With the latest commits I have reverted to the previous scheme of terminating after just one level of recursion, but with the added twist that we track both the source and target sides (similarly to `recursiveTypeRelatedTo`) and terminate only when both have a circularity.

Intuitively, in inference we want to terminate when we encounter a duplicate attempt to infer from source and target types with the same origin, so `getRecursionIdentity` needs to get us as close as possible to the AST node that caused the type instantiation. Because most type instantiations are interned and shared (and thus have no reference to their originating AST node), this isn't always possible. For example, when inferring from `Box2<Box2<string>>` to `Box1<Box1<T>>`, where `Box1` and `Box2` are unique but structurally identical types, we end up with the same recursion identity for each `Box1` and `Box2` reference, and therefore terminate inference prematurely. This is a known problem (i.e. not new to this PR) and something we should continue to think about. A brute force way to work around the problem is to allow multiple levels of recursion, but that only works up to some level of nested and, as illustrated by the test failures, generates way too much work in general.

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test this

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -19543,7 +19515,24 @@ namespace ts {
...
+                // We stop inferring and report a circularity if we encounter duplicate recursion identities on both
+                // the source side and the target side.
+                const saveExpandingFlags = expandingFlags;
+                const sourceIdentity = getRecursionIdentity(source);
+                const targetIdentity = getRecursionIdentity(target);
+                if (sourceIdentity && contains(sourceStack, sourceIdentity)) expandingFlags |= ExpandingFlags.Source;
+                if (targetIdentity && contains(targetStack, targetIdentity)) expandingFlags |= ExpandingFlags.Target;
+                if (expandingFlags !== ExpandingFlags.Both) {
+                    if (sourceIdentity) (sourceStack || (sourceStack = [])).push(sourceIdentity);
+                    if (targetIdentity) (targetStack || (targetStack = [])).push(targetIdentity);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:19527]:
```suggestion
                    if (sourceIdentity) (sourceStack ||= []).push(sourceIdentity);
                    if (targetIdentity) (targetStack ||= []).push(targetIdentity);
```

**Comment by AlCalzone** [comment]
@ahejlsberg According to your example in the OP, this also fixes #26223 😊

**Comment by ahejlsberg** [comment]
Test runs all look clean. Slight regression in check time for material-ui, but it's worth it for the added precision in type inference.

File: src/compiler/checker.ts (hunk 3 above)
Comment (ahejlsberg) [reply src/compiler/checker.ts:19527]: I'm going to leave it for now. We have *lots* of occurrences of that pattern, so maybe another PR to clean them all up.

File: src/compiler/checker.ts (hunk 2 above)
Comment (ahejlsberg) [reply src/compiler/checker.ts:18224]: No. Previously conditional types couldn't be recursive, so they weren't included here. But for that same reason we have no tests that could be affected by this.

**Comment by strelga** [comment]
@ahejlsberg @rbuckton 

Does this change mean we no longer need hacks like `awaited` keyword to handle the recursive nature of `Promise`?

**Comment by tjjfvi** [comment]
Can we get a playground for this PR? I'd like to play around with the new options this gives us.

**Comment by levenleven** [comment]
@tjjfvi you can choose "Nightly" version 
image

**Comment by groege** [comment]
Using this PR I'm finally able to to a lot of things not working before.
The biggest thing was a library (kendo-ui) which wraps classes and on can use getters and setters in the following way:
```
class Test {
    PropertyA!: string;
    PropertyB!: Test;
}

var test = new Test();
test.PropertyB = new Test();
test.PropertyB.PropertyB = new Test();
var wrapped = kendo.observable(test );
wrapped.set('PropertyB.PropertyB.PropertyA', 'test');
```

Using this PR I was able to implement a Type that lists all possibilities, which obviously would fail because it is circular because it could go endlessly PropertyB.PropertyB.PropertyB..........

I've tried fixed it the following way
```
    type PathExcludes = '__typename';
    type PathImpl<T, TFound extends string, TPref extends keyof T = never> =
        TFound extends `${infer _}.${TFound}.${infer _}` ? never :
        TPref extends never
        ? PathImpl<T, TFound, keyof T> | keyof T
        : TPref extends string
        ? T[TPref] extends string | number | Date ? never
        : `${TPref}.${PathImpl<T[TPref], `..${T & string}..`, Exclude<keyof T[TPref], PathExcludes>> & string}` | `${TPref}.${Exclude<keyof T[TPref], PathExcludes> & string}`
        : keyof T;
    type PathImpl2<T> = PathImpl<T, ``, Exclude<keyof T, PathExcludes>> | Exclude<keyof T, PathExcludes>;
    type Path<T> = (PathImpl2<T> extends string | keyof T ? PathImpl2<T> : keyof T);
```

See also [playground](https://www.typescriptlang.org/play?ts=4.1.0-dev.20200921#code/MYGwhgzhAEAqCmEAu0DeAoaXoAUBOA9gA7x5ICeAggIQBcciSATANybb7GkUBCdDyNgF906UJBgJkTNOyycSZKvynNho7NAolcYJAAsAogA9QAVwAmiaAF5oAcgD6j7fAB2YALbx7bTa90DAElPIhAAHlgAGjgAMQIzNwtoeGMkdwsYZDwASzcAcxjYfHgAMxS0jJgAa3hyAnLYW2g3eAA3UgA+WzlNOISkivSkmAADABJUPNLSaEchADpJ2HjEi0XJ6dn50egAfhb22dpezWK8MqGqw468U+wDnD19ELDIotWkmNr6xu6AH2gPwacHuWHo50uqWGmWg2Ty+TB+zgAG1IaUALpXEZwpC5ArQQFuMyeABGs0BABE9PBka1bkj6BNUOiNqgnsFQhFYGiSpiYqMFksWdAAGS4-H5RYLUYxEzmKzhYGNXkXfmBIymECWRCdbri+EFIS7QHM1nC+XaxXK1HojExDmahW6sUShHGxlAuog2B+bABR2vEBMSLdOyBrnvaCjWXQS06pXexoO57xqwQPWEuNahM26IatO6v1YAPPUPNAAUEbCIdg3Wh10N+SzeeR1eD5foeYAlGxRBpsG0wHhoMZ6I7IoxOsXR3Z7L4gA)

Is there anyway to fix and simplify this? As of now it will only show one child...
And even better to allow to give a hint to the type so it will only load the next X number of possibilities?
This would actually allow

**Comment by lusess123** [comment]
> ```ts
> type T4 = TupleOf<number, 100>;  // Depth error
> ```

why the max number is 43 ? not 48 ?

**Pull request merged**